s3repo/model.py

```diff
@@ -180,13 +181,137 @@ def _upload_files(self, package, tarantool_series, origin_files):
+        # (it is enough for us to know that the first level content is exists).
+        #
+        # See https://github.com/boto/boto3/issues/134 for how to list first
+        # level content by a specific prefix.
+
+        path = base_path + '/'
+        list_objs = self.bucket.meta.client.list_objects_v2(
+            Bucket=self.bucket.name,
+            Delimiter='/',
+            Prefix=path
+        )
+        if list_objs.get('CommonPrefixes') is None:
+            return []
+
+        # In the case of a deb-base distribution, the meta information about
+        # packages in all versions of the distribution is updated together.
+        # So we just return the path to the distribution
+        return [path]
+
+    def _get_rpm_repo_path(self, base_path, dist_versions):
+        """Returns the list of the paths to the rpm-based reposies
+        for updating metainformation with the 'mkrepo' tool.
+        base_path(string) - path to the distribution.
+        dist_versions(list) - list of the distribution versions.
+        """
+        # See the first comment in "_get_deb_repo_path" method.
+
+        # In the case of an rpm-based distribution, one distribution version can
+        # include several repositories (for example "x86_64" and "SRPM").
+        # We must collect them all.
+        repos_list = []
+        for ver in dist_versions:
+            # Path to all repositories of the distribution version.
+            common_path = '/'.join([base_path, ver]) + '/'
+            list_objs = self.bucket.meta.client.list_objects_v2(
+                Bucket=self.bucket.name,
+                Delimiter='/',
+                Prefix=common_path
+            )
+            dist_repos_list = list_objs.get('CommonPrefixes')
+            if dist_repos_list is None:
+                continue
+            for repo in dist_repos_list:
+                # In this context, "Prefix" is a path to the repository.
+                repos_list.append(repo['Prefix'])
+
+        return repos_list
+
+    def _get_repository_list(self):
+        """Returns a list of paths to repositories in the current bucket."""
+        supported_repos = self.get_supported_repos()
+        repos_list = []
+        result_list = []
+
+        # Since collecting the list of paths to repositories involves a large
+        # number of S3 requests through the network, it is recommended to use a
+        # thread pool to execute them in parallel (the thread will be idle for a
+        # "long" time, waiting for a response from S3).
+        #
+        # The number of processes = 20 was chosen experimentally.
+        with ThreadPool(processes=20) as pool:
+            for kind in supported_repos['repo_kind']:
+                for series in supported_repos['tarantool_series']:
+                    for dist, dist_description in supported_repos['distrs'].items():
+                        path = ''
+                        base_path = self.s3_settings.get('base_path', '')
```

Comment:
I'm looking into the surrounding code (nested for loops, paths formatting) and it looks as a good candidate to incapsulate into some class that will intellectually provide configured repositories information.

Sketchy code:

```python
for repo in config.repositories:
    process_path(repo.path)
    # repo.tarantool_series, repo.os, repo.dist, ...
```

Feel free to ignore, just raw idea.

Reply:
(I know, you don't use OS/DIST terminology here, but distr/dist_description['versions'] instead. However, to be honest, my ducking syndrome does not allow me to just bypass it (together with animus toward uncommon abbreviations). Sorry.)

Reply:
Debian, Ubuntu ... are not OS, they are distributions, and what you call a distribution are versions of a distribution(or releases) - https://en.wikipedia.org/wiki/Ubuntu
Why do you prefer to call them differently (OS and dist)?

Reply:
> I'm looking into the surrounding code (nested for loops, paths formatting) and it looks as a good candidate to incapsulate into some class that will intellectually provide configured repositories information.

I agree that this part of the code is not very good, but it is localized in a small (less than 50 lines) function.
After some reflection, I think that I do not understand your idea well, I will try to discuss it with you in my voice.

Reply:
We had a conversation with Alexander and AFAIU, such class is something like a repository manager that will work not with a configuration dictionary, but with some repos objects.
It is not related with current patchset, I will keep the idea for future changes.